Doxygen changes to usdGeom/mesh.h for review (work in progress)

barfowl opened the pull request:
As discussed, here are some potential documentation changes for UsdGeomMesh for review.

I just updated usdGeom/mesh.h for now to minimize the diffs and make the changes clearer -- I'll apply the final changes to usdGeom/schema.usda and regenrate mesh.h later when necessary.

Please add comments for any improvements, corrections or questions.

Thanks

barfowl committed:
Doxygen changes to usdGeom/mesh.h for review (work in progress):
    - reworded introduction to the Mesh class in general
    - added clarification of primvar interpolation types for meshes
    - improved descriptions for subd interpolation options
    - removed claims of watertightness between poly and bilinear meshes
    - mention of RI spec conflict with SHARPNESS_INFINITE constant
    - replaces http in links to https (all for OpenSubdiv)

barfowl commented on pxr/usd/usdGeom/mesh.h:
```diff
@@ -54,28 +54,69 @@ class SdfAssetPath;
+/// <b>A Note About Normals</b>
 /// 
-/// Normals should not be authored on a subdivided mesh, since subdivision
+/// Normals should not be authored on a subdividion mesh, since subdivision
```
Shoot, I had seen that spelling mistake earlier and thought I corrected it before the push...

c64kernal commented on pxr/usd/usdGeom/mesh.h:
```diff
@@ -54,28 +54,69 @@ class SdfAssetPath;
+/// as are associated terms such as "point index" and "vertex index", etc.
+/// The encoding defines the mesh topology completely in terms of a set of
+/// faces and their defining face-vertices (see _faceVertexCounts_ and
+/// _faceVertexIndices_).  No boundary representation or additional
```
Is there a way to make these link to the attribute docs below? (Just curious!)

barfowl replied:
Do we really want to?  Originally I did not include them, but I saw forward references to the "subdivisionScheme" attribute in the following paragraph and so followed the convention and emphasis.  Should we similarly link "subdivisionScheme"?

barfowl replied:
This has now been addressed merging Spiff's suggestions in the latest commit.

c64kernal commented on pxr/usd/usdGeom/mesh.h:
```diff
@@ -54,28 +54,69 @@ class SdfAssetPath;
+/// \anchor UsdGeom_Mesh_Primvars
+/// <b>A Note About Primvars</b>
+///
+/// The following lists clarifies the number of elements for and the
```
Typo: lists?

barfowl replied:
Noted -- and added to a pending commit.

c64kernal commented on pxr/usd/usdGeom/mesh.h:
```diff
@@ -231,15 +272,9 @@ class UsdGeomMesh : public UsdGeomPointBased
-    /// and faster to render, depending on renderer and render mode.)
+    /// "none" (i.e. a polygonal mesh with no subdivision).  Polygonal meshes
+    /// are typically lighterweight and faster to render, depending on renderer
+    /// and render mode.
```
I agree that we don't need to make a big distinction about it, but I do feel like there's something missing if we don't say anything at all about the distinction between polymeshes and bilinear. Perhaps one way to resolve this is to list the options in a list like we do above for interpolations, and below for face varying linear interpolation modes?

barfowl replied:
Agreed, it is a bit too lean.  I like the idea of the list, but "bilinear" will precede "none" in that list, and contrasting the two seems seems more awkward as a result.  How about appending the sentence:   "Use of "bilinear" will produce a similar shape to a polygonal mesh and may offer additional guarantees of watertightness and additional subdivision features (e.g. holes) but may also not respect authored normals."

barfowl replied:
I actually expanded "subdivisionScheme" to the list you suggested for other reasons (now in recent commit) and appended the sentence noted here.

barfowl commented on pxr/usd/usdGeom/mesh.h:
```diff
@@ -54,28 +54,69 @@ class SdfAssetPath;
-/// \em subdivisionScheme attribute.
+/// surfaces and simpler polygonal meshes. This is achieved by varying the
+/// _subdivisionScheme_ attribute, which is set to specify Catmull-Clark
+/// subdivision by default, so polygonal meshes must always be expliclty
```
Spelling of "explicitly" noted.

spiffmon commented on pxr/usd/usdGeom/mesh.h:
```diff
@@ -54,28 +54,69 @@ class SdfAssetPath;
 /// 
+/// Meshes are conceptually defined in terms of vertices, edges and faces.
+/// As a point-based primitive, the terms "vertex" and "point" are synonymous,
+/// as are associated terms such as "point index" and "vertex index", etc.
```
I'd like to hear @mkuruc's take on this, but in all my documentation, I've been trying to draw a *distinction* between point and vertex, which is like an instancing relationship.  `points` defines the "pool of points", and a `vertex` associates one of those points with an edge/face... we wind up having many more vertices than we have points, typically, which follows (and I think, helps to make sense out of) from the fact that len(points) != sum_i(faceVertexCounts) . And all of the attribute doc supports this interpretation.  So these statements contradict our other language.

There are other places below where you've equated points and vertices - I won't call them all out, but I'd request they all go.

Of course, if I'm the only one who thinks it's useful to draw this distinction, then let's talk!

spiffmon replied:
Maybe a clearer lexicon would be to just use `point` and `face vertex`, and not use simple `vertex` at all.  Using `vertex` when we have a `points` attribute is just confusing, witness all the places where you've parenthetically added it after you mention `points` ???

mkuruc replied:
I agree with Spiff about the importance of distinguishing between vertex and point.  (For example, summing up the faceVertexCounts vector will give you the total number of vertices, not the total number of points.

I've found the "point" vs. "face vertex" convention useful in the past to avoid ambiguity as to what "vertex" means.

barfowl replied:
Okay, I will definitely get rid of the equivalence between point and vertex and use "point" exclusively.  I also prefer to use the term face-vertex.

I find it difficult (requiring discipline) to not use the term "vertex" when talking about meshes, but will do so here if needed.  The term "vertex" has unfortunately become ambiguous in the industry.  I'm coming form the OpenSubdiv perspective with these changes -- wanting to try and align the descriptions of the subdivision features -- and vertex is synonymous with point in OpenSubdiv.

Aside from the obvious multiple places to deal with point vs vertex usage, I'm now wondering if these leading statements in the introduction should be completely removed or replaced with something more accurate... I'll replace it with something that emphasizes avoidance of "vertex" in place of "point" and "face-vertex".

barfowl replied:
I rephrased the opening paragraph in a recent commit to emphasize the use avoidance of the term "vertex" in favor "point" and "face-vertex", along with removing use of "vertex".  Let me know how that opening paragraph works for you now.

barfowl replied:
This should be addressed in the recent commits -- please comment on them if necessary.

spiffmon commented on pxr/usd/usdGeom/mesh.h:
```diff
@@ -54,28 +54,69 @@ class SdfAssetPath;
+/// \li <b>uniform</b> One element for each face of the mesh; elements are
+///     typically not interpolated but are inherited by other faces derived
+///     from a given face (via subdivision, tessellation, etc.).
+/// \li <b>varying</b> One element for each vertex or point of the mesh;
```
This is one place where the distinction between point and vertex pays off:  this interpolation mode is specified per-point, _not_ per-vertex... the RenderMan (adopted by us) `vertex` interpolation mode is unfortunate because it is also per-point.

barfowl replied:
Noted, thanks -- "vertex" now removed.

spiffmon commented on pxr/usd/usdGeom/mesh.h:
```diff
@@ -54,28 +54,69 @@ class SdfAssetPath;
+///
+/// Primvar interpolation types and related utilities are described more
+/// generally in \ref Usd_InterpolationVals.
+///
```
Thank you - it is so useful having this info here!

mkuruc replied:
Is it worth clarifying that the "bilinear" and "none" will be bilinearly interpolated for "vertex" interpolation?

Perhaps-- One element for each point of the mesh.  Interpolation of the elements matches the subdivision scheme's interpolation of the point data (smooth for 'catmull-clark' and 'loop', and linear for 'none' and 'bilinear').

barfowl replied:
I'm reluctant to use the word "bilinear" for interpolation when we can't guarantee that is the case.  For "none" there is no guarantee that a quad is bilinearly interpolated -- and in most cases, it isn't (quads are often first triangulated). I've also seen "bilinear" used generally that ignores the fact that it is never applied for Loop subdivision in any context, so I try to avoid it.

I like part of your suggestion about adding "interpolation of the point data" -- and retaining the use of "linear" -- but I prefer not to enumerate all subdivision cases inline.  It then needs updating if we add another scheme later. How about we describe the interpolation behavior more clearly in an expanded list for "subdivisionScheme"?

mkuruc replied:
Sounds great to me!

barfowl replied:
The new list form of "subdivisionScheme" is now included in the recent commit -- let me what you think.

barfowl replied:
No objections or comments so resolving

barfowl committed:
Revisions based on feedback:
    - fixed typos and other minor suggestions
    - reworked and separated "point" vs "vertex" paragraph in introduction
    - emphasized "point" and "face-vertex" in primvar interpolation
    - removed use of "vertex" in favor of "point" or "face-vertex"
    - expanded "subdivisionScheme" with list and properties per scheme
    - appended note on "bilinear" vs "none" for watertightness, normals, etc.

barfowl committed:
Merged suggestions about mesh topology based on feedback:
    - preserved but modified the avoidance of the term "vertex"
    - broke down the suggested topology description into more sentences
    - preserved the mention of lack of adjacency or neighborhood queries

barfowl committed:
Doc changes moved into schema.usda to generate mesh.h:
    - moved edits from mesh.h into schema.usda
    - used usdGenSchema to update mesh.h and generatedSchema.usda

barfowl committed:
Minor revisions for formatting consistency:
    - added wording to holeIndices for subdivision feature
    - minor rephrasing of cornerIndices and cornerSharpnesses
    - increased use of markdown in formatting and lists
    - use of single-quotes for introducing terminology
    - use of italics for attribute names
    - use of double-quotes for token names
    - use of bold label names in named lists
    - removal of Python namespaces

spiffmon commented on pxr/usd/usdGeom/mesh.h:
```diff
@@ -83,22 +83,22 @@ class SdfAssetPath;
 /// 
-/// \li <b>constant</b> One element for the entire mesh; no interpolation.
-/// \li <b>uniform</b> One element for each face of the mesh; elements are
+/// - __constant__: One element for the entire mesh; no interpolation.
```
Lovely - I forgot to answer your list-formatting question, but this is the preferred form!

spiffmon commented on pxr/usd/usdGeom/mesh.h:
```diff
@@ -228,7 +228,7 @@ class UsdGeomMesh : public UsdGeomPointBased
     // --------------------------------------------------------------------- //
-    /// Flat list of the index (into the 'points' attribute) of each
+    /// Flat list of the index (into the _points_ attribute) of each
     /// point of each face in the mesh.  If this attribute has more than
```
Given what we've said above, and the *name* of this attribute, would it not be more appropriate to use `vertex` here instead of `point` , since the vertex is already in reference to the face?

barfowl replied:
Agreed, I was trying to strictly avoid "vertex" -- except as the token for the interpolation type -- but it makes sense here.  Will include in another commit.

spiffmon commented on pxr/usd/usdGeom/mesh.h:
```diff
@@ -253,7 +253,7 @@ class UsdGeomMesh : public UsdGeomPointBased
     // FACEVERTEXCOUNTS 
     // --------------------------------------------------------------------- //
     /// Provides the number of points in each face of the mesh, 
```
Ditto to previous comment re: points vs vertices

barfowl replied:
Noted -- the combination of "face" and "vertex" clearly makes for "face-vertex", "face" and "point" do not.

spiffmon commented on pxr/usd/usdGeom/mesh.h:
```diff
@@ -355,23 +355,23 @@ class UsdGeomMesh : public UsdGeomPointBased
     // --------------------------------------------------------------------- //
-    /// Specifies how elements of a _faceVarying_ primvar are
-    /// interpolated. Interpolation can be as smooth as a _vertex_ primvar or
+    /// Specifies how elements of a "faceVarying" primvar are
```
In the intro, you used single-quotes for 'points' etc... so why not single-quotes here, too?

barfowl replied:
The single-quotes were to introduce a general term.  Here the references is to specific token names for the primvar interpolation type -- and double-quotes are used for token names.  I can see that using it as an adjective might obscure that...  Perhaps "how elements of a primvar of interpolation type "faceVarying"..." is warranted...

barfowl replied:
Rephrased as suggested in subsequent commit

barfowl committed:
Minor changes in response to comments:
    - replaced use of "points" with "vertices" in context of face-vertices
    - clarified primvar interpolatin type with double-quoted token name
    - changed "discontinuous" boundaries to "interior" in fvar interpolation

barfowl committed:
Improved descriptions of primvar interpolation type tokens.

barfowl commented:
Changes here have now been merged in the main repository (with the exception of the last commit affecting primvar interpolation tokens), so we can close this.  Thanks for all the helpful feedback.
